app/code/Magento/Wishlist/Helper/Data.php

```diff
@@ -630,4 +639,36 @@ public function getProductUrl($item, $additional = [])
         }
         return $product->getUrlModel()->getUrl($product, $additional);
     }
+
+	/**
+     * Refresh and set new form_key in session and cookie
+     *
+     * @return $this
+     */
+    public function refreshFormKey()
+    {
+        $beforeWishlistRequest = $this->_customerSession->getBeforeWishlistRequest();
+
+        if ($beforeWishlistRequest !== null) {
+
+            //form_key is already empty
+            $formKey = $this->fromKey->getFormKey();
+
+            $cookieMetadata = $this->cookieMetadata->createPublicCookieMetadata();
+            $cookieMetadata->setDomain($this->configInterface->getCookieDomain());
+			$cookieMetadata->setPath($this->configInterface->getCookiePath());
+            $cookieMetadata->setDuration($this->configInterface->getCookieLifetime());
+
+            $this->pageCacheForm->set($formKey,$cookieMetadata);
```

Comment:
Looks like this logic eliminates the whole point in form key, isn't it? As after such change CSRF is possible with automatic generation of form key.

Reply:
@orlangur Reason of setting form key is to take selected product from guest session to customer session, when user logs in.